chai_lab/data/io/cif_utils.py

```diff
@@ -39,60 +35,88 @@ class _LocalPLDDT(modelcif.qa_metric.Local, modelcif.qa_metric.PLDDT):
+) -> list[str]:
+    return [
+        "".join([chr(ord_val + 32) for ord_val in ords_atom]).rstrip()
+        for ords_atom in tensor
+    ]
+
+
+@typecheck
 def token_centre_plddts(
     context: PDBContext,
+    plddts: Float[Tensor, "n"],
     asym_id: int,
 ) -> tuple[list[float], list[int]]:
-    assert context.atom_bfactor_or_plddt is not None
+    assert plddts is not None
     mask = context.token_asym_id == asym_id
 
     atom_idces = context.token_centre_atom_index[mask]
 
     # residue indices for tokens
     residue_indices = context.token_residue_index[mask]
 
-    return context.atom_bfactor_or_plddt[atom_idces].tolist(), residue_indices.tolist()
+    return plddts[atom_idces].tolist(), residue_indices.tolist()
 
 
-def get_chains_metadata(context: PDBContext) -> list[dict]:
+def get_chains_metadata(context: PDBContext, entity_names) -> dict[int, AsymUnit]:
     # for each chain, get chain id, entity id, full sequence
     token_res_names = context.token_res_names_to_string
 
-    records = []
+    asym_id2asym_unit = {}
 
-    for asym_id in torch.unique(context.token_asym_id):
-        if asym_id == 0:  # padding
-            continue
+    for asym_id, entity_type in context.asym_id2entity_type.items():
+        assert asym_id != 0, "zero is padding for asym_id"
 
-        token_indices = torch.where(context.token_asym_id == asym_id)[0]
+        [token_indices] = torch.where(context.token_asym_id == asym_id)
         chain_token_res_names = [token_res_names[i] for i in token_indices]
 
         residue_indices = context.token_residue_index[token_indices]
 
         # check no missing residues
-        max_res = torch.max(residue_indices).item()
-        tokens_in_residue = (
-            residue_indices == torch.arange(max_res + 1)[..., None]
-        )  # (residues, tokens)
-        assert tokens_in_residue.any(dim=-1).all()
-
-        first_token_in_resi = torch.argmax(tokens_in_residue.int(), dim=-1)
-
-        sequence = [chain_token_res_names[i] for i in first_token_in_resi]
-        entity_id = context.token_entity_id[token_indices][0]
-
-        entity_type: int = context.get_chain_entity_type(asym_id)
-
-        records.append(
-            {
-                "sequence": sequence,
-                "entity_id": entity_id.item(),
-                "asym_id": asym_id.item(),
-                "entity_type": entity_type,
-            }
+        max_res = int(torch.max(residue_indices).item())
+
+        any_token_in_resi = residue_indices.new_zeros([max_res + 1]) - 99
+
+        any_token_in_resi[residue_indices] = torch.arange(
+            len(residue_indices),
+            dtype=residue_indices.dtype,
+            device=residue_indices.device,
+        )
+
+        assert any_token_in_resi.min() >= 0
+
+        sequence = [chain_token_res_names[i] for i in any_token_in_resi]
+
+        # TODO confirm we want to use entity_id not asym_id.
+        entity_id = context.token_entity_id[token_indices[0]]
```

Comment:
I'm fairly sure that we do want `entity_id` here because we are creating the asym unit records, though worth checking that this works properly with a homo-dimer example.